LCOW: CLI changes to add platform flag - pull, run, create and build

jhowardmsft opened the pull request:
Signed-off-by: John Howard jhoward@microsoft.com

Replacement for https://github.com/docker/cli/pull/416.  This adds the CLI flag changes to match those described in https://github.com/moby/moby/issues/34617, and PR'd to moby/moby in https://github.com/moby/moby/pull/34642.

It adds `--platform` to `pull`, `run`, `create`, and `build`. Note `import` requires further daemon changes so has not been done yet (will be in a further PR once the daemon has been updated).

dnephin commented on cli/command/commands/commands.go:
```diff
@@ -87,6 +87,7 @@ func AddCommands(cmd *cobra.Command, dockerCli *command.DockerCli) {
 		hide(container.NewLogsCommand(dockerCli)),
 		hide(container.NewPauseCommand(dockerCli)),
 		hide(container.NewPortCommand(dockerCli)),
+		hide(image.NewPullCommand(dockerCli)),
```
looks like this moved for no reason? now it's not sorted

dnephin commented on cli/command/container/run.go:
```diff
@@ -123,6 +128,7 @@ func runContainer(dockerCli *command.DockerCli, opts *runOptions, copts *contain
 	hostConfig := containerConfig.HostConfig
 	stdout, stderr := dockerCli.Out(), dockerCli.Err()
 	client := dockerCli.Client()
+	ociPlatform := system.ParsePlatform(opts.platform)
```
I think this should be parsed by the flag instead after flag parsing. To accomplish this you can create a new `PlatformOpt` type in `opts/platform.go` with a `func Value() spec.Platform` to return the parsed type.

The field in the options struct will be a `PlatformOpt` instead of a `string`

dnephin commented on cli/command/container/run.go:
```diff
@@ -59,6 +61,9 @@ func NewRunCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.StringVar(&opts.detachKeys, "detach-keys", "", "Override the key sequence for detaching a container")
 
+	flags.StringVar(&opts.platform, "platform", "", "Set platform if server is multi-platform capable")
+	flags.SetAnnotation("platform", "version", []string{"1.32"})
```
Since this is going to be repeated in a bunch of commands how about we move this to 

```go
func addPlatformFlag(flags *pflag.FlagSet, opt *PlatformOpt)
```

in `cli/flags/common.go`


There will also need to be another annotation for hiding based on the header:

```go
flags.SetAnnotation("platform", "platform", nil)
```

The code to hide flags based on annotations is in `cmd/docker/docker.go`

dnephin commented on cli/command/image/cmd.go:
```diff
@@ -20,8 +20,8 @@ func NewImageCommand(dockerCli command.Cli) *cobra.Command {
 		NewLoadCommand(dockerCli),
-		NewPullCommand(dockerCli),
 		NewPushCommand(dockerCli),
+		NewPullCommand(dockerCli),
```
unsorted

dnephin commented on cli/command/image/trust.go:
```diff
@@ -296,7 +298,10 @@ func trustedPull(ctx context.Context, cli command.Cli, repoInfo *registry.Reposi
-func imagePullPrivileged(ctx context.Context, cli command.Cli, authConfig types.AuthConfig, ref string, requestPrivilege types.RequestPrivilegeFunc, all bool) error {
+func imagePullPrivileged(ctx context.Context, cli command.Cli, authConfig types.AuthConfig, ref string, requestPrivilege types.RequestPrivilegeFunc, all bool, platform specs.Platform) error {
+	if platform.OS == "" && os.Getenv("DOCKER_DEFAULT_PLATFORM") != "" {
+		platform.OS = os.Getenv("DOCKER_DEFAULT_PLATFORM")
```
This should be handled somewhere else before it gets into the pull handling. Possibly as part of the `PlatformOpt` type. It can be used as a default value for `Value()`

jhowardmsft commented:
Rebased and reworked based on https://github.com/moby/moby/pull/34642 and addressed most of the comments. Still a work in progress until 34642 is merged in moby/moby as the vendoring will remain hacked.

jhowardmsft commented:
Rebased to address conflicts, and reworked to use two utility functions rather than replicate the same code in a few places.

Ready for review, removed WIP from the title, updated the commit comment and the description of the PR.  @vdemeester can this be moved to code review now?

jhowardmsft commented:
@johnstep PTAL too. Thanks.

dnephin commented on cli/command/utils.go:
```diff
@@ -117,3 +118,18 @@ func PruneFilters(dockerCli Cli, pruneFilters filters.Args) filters.Args {
+
+// PlatformFromEnvironment sets the platform from an environment variable
+// if it isn't otherwise populated.
+func PlatformFromEnvironment(platform string) string {
```
Since this is effectively setting a default value for platform I think the environment variable should be read as the default value in `AddPlatformFlag()` instead of exporting this function:

```go
flags.StringVar(target, "platform", os.Getenv("DOCKER_DEFAULT_PLATFORM"), ...)
```

jhowardmsft replied:
That makes sense. Updated.

For hiding the flag, it is going to require getting consensus on https://github.com/moby/moby/pull/34628 first, which I don't think there is among other maintainers yet.

johnstep commented:
This should probably be hidden for a daemon without experimental.

jhowardmsft commented:
@johnstep - OK, added that annotation in utils.go::AddPlatformFlag() and verified it operates as expected.

jhowardmsft commented:
Is this OK to merge with 2 LGTM's?
